File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-0.java
class Solution {
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
      return helper(root, subRoot);
    }

    private boolean helper(TreeNode node, TreeNode subRoot) {
      if (node == null) return false;

      if (!helperHelper(node, subRoot)) {
        return (helper(node.left, subRoot) || helper(node.right, subRoot));
      }

      return true;
    }

    private boolean helperHelper(TreeNode node, TreeNode subRoot) {
      if (node == null && subRoot != null) return false;
      if (node != null && subRoot == null) return false;
      if (node == null && subRoot == null) return true;

      if (node.val == subRoot.val)
        return (
          helperHelper(node.left, subRoot.left) &&
          helperHelper(node.right, subRoot.right)
        );
      return false;
    }
}

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.cpp

class Solution {
public:
    int total = 0;

    int goodNodes(TreeNode* root) {
      helper(root, root->val);

      return total;
    }

    void helper(TreeNode* node, int maxSoFar) {
      if (!node) return;
      if (node->val >= maxSoFar) {
        total++;
        maxSoFar = node->val;
      }

      helper(node->right, maxSoFar);
      helper(node->left, maxSoFar);
    }
};
